toponetx/classes/path.py

```diff
@@ -0,0 +1,162 @@
+"""Path class."""
+
+from collections.abc import Hashable, Iterable, Sequence
+from typing import Any
+
+from toponetx.classes.complex import Atom
+
+__all__ = ["Path"]
+
+
+class Path(Atom):
+    """
+    A class representing a p-path in a path complex, which is the building block for a path complex.
+
+    Unlike the original paper (https://arxiv.org/pdf/1207.2834.pdf) where p-paths span the regular space of boundary-invariant paths,
+    our p-paths span the space of simple paths with length p.
+
+    Parameters
+    ----------
+    elements: Sequence[Hashable]
+        The nodes in the p-path.
+    name : str, optional
+        A name for the p-path.
+    construct_boundaries : bool, default=False
+        If True, construct the entire boundary of the p-path.
+    reserve_sequence_order : bool, default=False
+        If True, reserve the order of the sub-sequence of nodes in the p-path.
+        Else, the sub-sequence of nodes in the p-path will be reversed if the first index is larger than the last index.
+    allowed_paths : Iterable[tuple[Hashable]], optional
+        A list of allowed boundaries. If None, all possible boundaries are constructed (similarly to simplex).
+    attr: keyword arguments, optional
+        Additional attributes to be associated with the p-path.
+    """
+
+    def __init__(
+        self,
+        elements: Sequence[Hashable],
+        name: str = "",
+        construct_boundaries: bool = False,
+        reserve_sequence_order: bool = False,
+        allowed_paths: Iterable[tuple[Hashable]] = None,
+        **attr,
+    ) -> None:
+        self.__check_inputs(elements, reserve_sequence_order)
+        super().__init__(tuple(elements), name, **attr)
+        if len(set(elements)) != len(self.elements):
+            raise ValueError("A p-path cannot contain duplicate nodes.")
+
+        self.construct_boundaries = construct_boundaries
+        self.reserve_sequence_order = reserve_sequence_order
+        if construct_boundaries:
+            self._boundaries = self.construct_path_boundaries(
+                elements,
+                reserve_sequence_order=reserve_sequence_order,
+                allowed_paths=allowed_paths,
+            )
+        else:
+            self._boundaries = list()
+
+    @staticmethod
+    def construct_path_boundaries(
+        elements: Sequence[Hashable],
+        reserve_sequence_order: bool = False,
+        allowed_paths: Iterable[tuple[Hashable]] = None,
+    ) -> list[tuple[Hashable]]:
+        """
```

Comment:
define this method mathematically, give an example. Make it easy and pythonic.

Reply:
Just added.